File: Deque/src/Deque.java

public class Deque {
	
	private class Node {
		public int value;
		public Node next;
		
		public Node(int v, Node n) {
			value = v;
			next = n;
		}
	}
	
	private Node head;
	private Node tail;
	
	public boolean isEmpty() {
		return head == null;
	}
	
	public void print() {
		Node n = head;
		while (n != tail) {
			System.out.println(n.value);
			n = n.next;
		}
		System.out.println(n.value);
	}
	
	public void insertElement(int value, boolean fromHead) {
		if (isEmpty()) {
			tail = head = new Node(value, null);
		} else {
			if (fromHead) {
				head = new Node(value, head);
			} else {
				tail.next = new Node(value, null);
				tail = tail.next;
			}
		}
	}
	
	public void deleteElement(boolean fromHead) {
		if (!isEmpty()) {
			if (fromHead) {
				head = head.next;
			} else {
				Node n = head;
				while (n.next != tail) 
					n = n.next;
				n.next = null;
				tail = n;
			}
		}
	}
	
	public int getElement(boolean fromHead) {
		if (!isEmpty()) {
			if (fromHead) {
				return head.value;
			} else {
				return tail.value;
			}
		}
		
		return 0;
	}
}

File: Deque_C/deque.cpp
#include "deque.hpp"

Deque::Deque()
{
	head = NULL;
	tail = NULL;
};

Deque::~Deque()
{
	node* n;
	while (head != tail)
	{
		n = head;
		head = head->next;
		delete n;
	};
};

bool Deque::isEmpty()
{
	return head == NULL;
};

void Deque::print()
{
	node* n = head;
	while (n != tail)
	{
		std::cout << n->value << " ";
		n = n->next;
	};
	std::cout << n->value << std::endl;
};

void Deque::insertElement(int value, bool from_head)
{
	if (isEmpty()) 
	{
		head = tail = new node(value);
	} 
	else
	{
		if (from_head) {
			head = new node(value, head);
			//insertElement(value, from_head, head);
		}
		else
		{	
			tail->next = new node(value);
			tail = tail->next;
			//insertElement(value, from_head, tail);
		}
	}
};

void Deque::deleteElement(bool from_head)
{
	if (!isEmpty())
	{
		if (from_head)
		{
			node* n = head;		
			head = head->next;
			delete n;
		}
		else
		{	
			node* n = head;
			while (n->next != tail)
				n = n->next;
			delete tail;
			n->next = NULL;
			tail = n;
		}
	}
}; 

int Deque::getElement(bool from_head)
{
	if (!isEmpty())
	{
		if (from_head)
		{
			return head->value;
		}
		else
		{
			return tail->value;
		}
	}
	
	return -1;
};
